src/devices/Hmc5883l/Hmc5883l.cs

```diff
@@ -0,0 +1,119 @@
+        public Vector3 DirectionVector => ReadDirectionVector();
+
+        /// <summary>
+        /// HMC5883L Heading (DEG)
+        /// </summary>
+        public double Heading => VectorToHeading(ReadDirectionVector());
+
+        /// <summary>
+        /// Initialize a new HMC5883L device connected through I2C
+        /// </summary>
+        /// <param name="sensor">I2C Device, like UnixI2cDevice or Windows10I2cDevice</param>
+        /// <param name="gain">Gain Setting</param>
+        /// <param name="measuringMode">The Mode of Measuring</param>
+        /// <param name="outputRate">Typical Data Output Rate (Hz)</param>
+        public Hmc5883l(I2cDevice sensor, Gain gain = Gain.Gain1090, MeasuringMode measuringMode = MeasuringMode.Continuous, OutputRate outputRate = OutputRate.Rate15)
+        {
+            _sensor = sensor;
+            _gain = (byte)gain;
+            _measuringMode = (byte)measuringMode;
+            _outputRate = (byte)outputRate;
+
+            Initialize();
+        }
+
+        /// <summary>
+        /// Initialize the sensor
+        /// </summary>
+        private void Initialize()
+        {
+            // Details in Datasheet P12
+            byte configA = (byte)(0x70 + _outputRate << 2);
+            byte configB = (byte)(_gain << 5);
+
+            Span<byte> commandA = stackalloc byte[] { (byte)Register.HMC_CONFIG_REG_A_ADDR, configA };
+            Span<byte> commandB = stackalloc byte[] { (byte)Register.HMC_CONFIG_REG_B_ADDR, configB };
+            Span<byte> commandMode = stackalloc byte[] { (byte)Register.HMC_MODE_REG_ADDR, _measuringMode };
+
+            _sensor.Write(commandA);
+            _sensor.Write(commandB);
+            _sensor.Write(commandMode);
+        }
+
+        /// <summary>
+        /// Read raw data from HMC5883L
+        /// </summary>
+        /// <returns>Raw Data</returns>
+        private Vector3 ReadDirectionVector()
+        {
+            Span<byte> xRead = stackalloc byte[2];
+            Span<byte> yRead = stackalloc byte[2];
+            Span<byte> zRead = stackalloc byte[2];
+
+            _sensor.WriteByte((byte)Register.HMC_X_MSB_REG_ADDR);
+            _sensor.Read(xRead);
+            _sensor.WriteByte((byte)Register.HMC_Y_MSB_REG_ADDR);
+            _sensor.Read(yRead);
+            _sensor.WriteByte((byte)Register.HMC_Z_MSB_REG_ADDR);
+            _sensor.Read(zRead);
+
+            short x = BinaryPrimitives.ReadInt16BigEndian(xRead);
+            short y = BinaryPrimitives.ReadInt16BigEndian(yRead);
+            short z = BinaryPrimitives.ReadInt16BigEndian(zRead);
+
+            return new Vector3(x, y, z);
+        }
+
+        /// <summary>
+        /// Calculate heading
+        /// </summary>
+        /// <param name="vector">HMC5883L Direction Vector</param>
+        /// <returns>Heading (DEG)</returns>
+        private double VectorToHeading(Vector3 vector)
+        {
+            return Math.Atan2(vector.Y, vector.X) * (180 / Math.PI) + 180;
```

Comment:
```csharp
double ret = Math.Atan2(vector.Y, vector.X) * 180 / Math.PI;
if (ret < 0)
  ret += 360;
return ret;
```

after that LGTM

Reply:
just to clarify: +180 reverses direction. Adding 360 points you to the same place and only changes the range of values

Reply:
@krwq  When I used the above algorithm, I got the wrong angle.
![wrong](https://user-images.githubusercontent.com/15957347/52613178-dfd29400-2ec7-11e9-82ef-5a34435d3ddf.jpg)
The correct results are as follows
![right](https://user-images.githubusercontent.com/15957347/52613179-dfd29400-2ec7-11e9-8bc2-d1ab88f2f2b8.jpg)

Reply:
To add, it's not a math problem, because that's how sensor works.

Reply:
@ZhangGaoxing I know one thing, adding 180 is not correct from mathematical point of view assuming the direction is pointing north. My suspicions are following:
- sensor is upside down
- the angle should not be measured on Z axis and on other instead
- direction is not showing north and rather it is showing exactly opposite

Here is how we can verify this:
- put sensor in couple of positions
- for each position show me the XYZ vector (all 3 components, no angle)

positions:
- flat on the table (horizontally) so that X showed on page 4 of datasheet points north
- same but so that Y points north
- vertically X should point up or down (write down which one it is) 
and you should see component exactly as shown on the page so that if you looked through it you would see north

We theoretically should get (1,0,0); (0,1,0); (0,0,1/-1) but I'm suspecting you'll get something else :-)

Reply:
basically whenever you point it to north one of the components should be close to 1 and every other should be close to 0 (it might be scaled so i.e. 0.33,0,0 would be acceptable too) - you can also normalize the vector if you get such results but it should not affect the 0 values

Reply:
You are right, my sensor is upside down. 
I'm truly sorry. My job is to do ASP.NET, writing hardware is entirely due to interest. I may not be thinking about a comprehensive problem, maybe the running results correctly cover up the code errors. After all, I am not professional. I have learned a lot from you and other members. Thanks again!

Reply:
@ZhangGaoxing thank you for contributing :smile: you seem to have more hardware knowledge than many of us, I just have some robotics background but not that much in hardware :smile: